src/Lean/Compiler/IR/EmitLLVM.lean

```diff
@@ -0,0 +1,1521 @@
+  LLVM.i64Type llvmctx
+end LLVM
+
+namespace EmitLLVM
+
+structure Context (llvmctx : LLVM.Context) where
+  env        : Environment
+  modName    : Name
+  jpMap      : JPParamsMap := {}
+  mainFn     : FunId := default
+  mainParams : Array Param := #[]
+  llvmmodule : LLVM.Module llvmctx
+
+structure State (llvmctx : LLVM.Context) where
+  var2val : HashMap VarId (LLVM.LLVMType llvmctx × LLVM.Value llvmctx)
+  jp2bb   : HashMap JoinPointId (LLVM.BasicBlock llvmctx)
+
+inductive Error where
+| unknownDeclaration : Name → Error
+| invalidExportName : Name → Error
+| unimplemented : String → Error
+| compileError : String → Error -- TODO: change this into an exhaustive Inductive enumeration.
+
+instance : ToString Error where
+  toString e := match e with
+   | .unknownDeclaration n => s!"unknown declaration '{n}'"
+   | .invalidExportName n => s!"invalid export name '{n}'"
+   | .unimplemented s => s!"unimplemented '{s}'"
+   | .compileError s => s!"compile error '{s}'"
+
+abbrev M (llvmctx : LLVM.Context) :=
+  StateT (State llvmctx) (ReaderT (Context llvmctx) (ExceptT Error IO))
+
+instance : Inhabited (M llvmctx α) where
+  default := throw (Error.compileError "inhabitant")
+
+def addVartoState (x : VarId) (v : LLVM.Value llvmctx) (ty : LLVM.LLVMType llvmctx) : M llvmctx Unit := do
+  modify (fun s => { s with var2val := s.var2val.insert x (ty, v) }) -- add new variable
+
+def addJpTostate (jp : JoinPointId) (bb : LLVM.BasicBlock llvmctx) : M llvmctx Unit :=
+  modify (fun s => { s with jp2bb := s.jp2bb.insert jp bb })
+
+def emitJp (jp : JoinPointId) : M llvmctx (LLVM.BasicBlock llvmctx) := do
+  let state ← get
+  match state.jp2bb.find? jp with
+  | .some bb => return bb
+  | .none => throw (Error.compileError s!"unable to find join point {jp}")
+
+def getLLVMModule : M llvmctx (LLVM.Module llvmctx) := Context.llvmmodule <$> read
+
+def getEnv : M llvmctx Environment := Context.env <$> read
+
+def getModName : M llvmctx  Name := Context.modName <$> read
+
+def getDecl (n : Name) : M llvmctx Decl := do
+  let env ← getEnv
+  match findEnvDecl env n with
+  | some d => pure d
+  | none   => IO.eprintln "getDecl failed!"; throw (Error.unknownDeclaration n)
+
+def constIntUnsigned (n : Nat) : M llvmctx (LLVM.Value llvmctx) :=  do
+    LLVM.constIntUnsigned llvmctx (UInt64.ofNat n)
+
+def getOrCreateFunctionPrototype (mod : LLVM.Module llvmctx)
+  (retty : LLVM.LLVMType llvmctx) (name : String) (args : Array (LLVM.LLVMType llvmctx)) : M llvmctx  (LLVM.Value llvmctx) := do
+  LLVM.getOrAddFunction mod name $
+     (← LLVM.functionType retty args (isVarArg := false))
+
+def callLeanBox (builder : LLVM.Builder llvmctx) (arg : LLVM.Value llvmctx) (name : String := "") : M llvmctx (LLVM.Value llvmctx) := do
+  let fnName :=  "lean_box"
+  let retty ← LLVM.voidPtrType llvmctx
+  let argtys := #[ ← LLVM.size_tType llvmctx ]
+  let fn ← getOrCreateFunctionPrototype (← getLLVMModule) retty fnName argtys
```

Comment:
It's not as bad as it looks, as [LLVM `Type`s are uniqued](https://llvm.org/doxygen/classllvm_1_1Type.html#details), so most of the time is spent in the fast path, looking up caches. 

Here's [the implementation of `FunctionType::get`](https://llvm.org/doxygen/Type_8cpp_source.html#l00360
) for example:
```cpp
// This is the factory function for the FunctionType class.
FunctionType *FunctionType::get(Type *ReturnType,
                                 ArrayRef<Type*> Params, bool isVarArg) {
   ...
   // Since we only want to allocate a fresh function type in case none is found
   // and we don't want to perform two lookups (one for checking if existent and
   // one for inserting the newly allocated one), here we instead lookup based on
   // Key and update the reference to the function type in-place to a newly
   // allocated one if not found.
   auto Insertion = pImpl->FunctionTypes.insert_as(nullptr, Key);
   if (Insertion.second) {
     // The function type was not found. Allocate one and update FunctionTypes
     // in-place.
     ...
  } else {
     // The function type was found. Just return it.
     FT = *Insertion.first;
   }
   return FT;
 }
```